### ISSUE TITLE
'cargo install' on local repo ignores Cargo.lock

### ISSUE THREAD

**Issue by ericseppanen** [rust-lang/cargo#9436]
**Problem**
I find that `cargo install` is persistently ignoring my `Cargo.lock` file: it always builds with different versions than saved in `Cargo.lock`. This causes rebuilding of many dependencies, wasting a lot of time and space.

**Steps**

1. `cargo init trythis; cd trythis`
2. add dependencies to `Cargo.toml`: note these are pinned to one version older than what's on crates.io.
```
[dependencies]
anyhow = "=1.0.39"
thiserror = "=1.0.23"
```
3. `cargo build`
```
    Updating crates.io index
   Compiling proc-macro2 v1.0.26
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.71
   Compiling anyhow v1.0.39
   Compiling quote v1.0.9
   Compiling thiserror-impl v1.0.23
   Compiling thiserror v1.0.23
   Compiling trythis v0.1.0 (/home/eric/work/rust/play/trythis)
    Finished dev [unoptimized + debuginfo] target(s) in 10.61s
```

4. Now that we have the older version in `Cargo.lock`, relax the dependencies in Cargo.toml (removing the `=`)
```
[dependencies]
anyhow = "1.0.39"
thiserror = "1.0.23"
```

5. Run `cargo build`; observe that it's doesn't rebuild because `Cargo.lock` still has us held to those older versions.

6. Run `cargo install --debug --path . --root /tmp/trythis`
```
  Installing trythis v0.1.0 (/home/eric/work/rust/play/trythis)
    Updating crates.io index
   Compiling anyhow v1.0.40
   Compiling thiserror-impl v1.0.24
   Compiling thiserror v1.0.24
   Compiling trythis v0.1.0 (/home/eric/work/rust/play/trythis)
    Finished dev [unoptimized + debuginfo] target(s) in 3.10s
  Installing /tmp/trythis/bin/trythis
   Installed package `trythis v0.1.0 (/home/eric/work/rust/play/trythis)` (executable `trythis`)
warning: be sure to add `/tmp/trythis/bin` to your PATH to be able to run the installed binaries
```

Observe:
- All dependencies get rebuilt, even though we just ran a build
- It ignores `Cargo.lock` and builds `thiserror v1.0.24` and `anyhow v1.0.40`.

**Notes**

Output of `cargo version`:
```
cargo 1.51.0 (43b129a20 2021-03-16)
```

**Comment by ericseppanen** [comment]
Hm, didn't realize this behavior is [deliberate](https://doc.rust-lang.org/cargo/commands/cargo-install.html):
> By default, the Cargo.lock file that is included with the package will be ignored. This means that Cargo will recompute which versions of dependencies to use, possibly using newer versions that have been released since the package was published. The --locked flag can be used to force Cargo to use the packaged Cargo.lock file if it is available.

This default seems questionable. Why would we want `cargo install` to produce binaries that are different from what `cargo build` produces? And why would that be left up to the user typing `cargo install`? A version-controlled `Cargo.lock` would seem to express the intent of the person publishing the crate; not honoring that seems unwise.

It's also quite annoying within CI pipelines, to discover that after carefully caching all of the dependencies, everything gets rebuilt anyway, because `--locked` wasn't specified.

**Comment by ericseppanen** [comment]
Also, the `cargo install -h` help text for `--locked` is quite confusing:
```
        --locked                    Require Cargo.lock is up to date
```
I have no idea what that means. What is "up to date"? What does "Require" mean? If I had to guess, I would have said that means "don't install this package unless every dependency in Cargo.lock is the latest available."

My first preference would be for `--locked` to be the default. My second choice would be for the help text to say `Don't ignore Cargo.lock` or `Honor dependency versions in Cargo.lock` or something like that.

**Comment by Eh2406** [comment]
The original implementation was motivated by the argument that we don't want to force all down stream binaries to release a new version to incorporate a security patch for all transitive dependencies. The fact that that was not entirely convincing is why we added the flag for the other behavior. The default is set to the original behavior, because we take backwards compatibility very seriously.

The help text, is not great. I would love a PR to help with it. Looks like it is using the same message for all commands, and on most commands [`--locked` does something different](https://doc.rust-lang.org/cargo/commands/cargo-build.html#manifest-options). Giving that the other meaning of the flag is available on basically all commands, I don't know haw easy it will be to customize it for `install`.

cc #7169 for some previous discussion and links.

**Comment by ericseppanen** [comment]
In hindsight,
> A version-controlled Cargo.lock would seem to express the intent of the person publishing the crate

seems a bit harsh.  After all, the crate author could have just as easily pinned the version in `Cargo.toml` if it was really important (i.e. some semver-violating breakage going on.)

I'm also sympathetic to both the security concerns and the need to preserve backwards compatibility.

In the end, I keep coming back to this: I just can't grasp how it's right to have `cargo build` and `cargo install` produce different outputs. This seems to violate basic principles of usability and discoverability.

If security patching is needed, it seems more sensible to ask the packager to delete the `Cargo.lock` file or pass in a `--update-dependencies` flag.

I would like to find a way to be more helpful than this. I am willing to attempt a PR to improve the help text... but that `build`/`install` disconnect just leaves me so baffled, that I have a hard time coming up with help text that makes sense.

Other thoughts:
- Perhaps the default behavior could be changed for the 2021 edition?
- If preserving the existing backwards-compatible behavior is important, perhaps it would be permissible to issue a warning? Maybe something like
`warning: a Cargo.lock file is present but will be ignored because --locked was not specified; some dependencies may be upgraded.`

**Comment by djc** [comment]
FWIW, I've also run into this issue and would be happy to contribute code if there is agreement from the team on how to approach fixing this.

(My preference would also be that cargo install start matching cargo build by default.)

**Comment by mqudsi** [comment]
This is a real problem for projects that publish working releases (as packaged binaries), have their code build and run fine on their dev machines, then have users complain that the code is broken when they check it out, build, and then try to install it via cargo from their locally checked out git directory.

**Comment by djc** [comment]
Should this be closed in favor of #7169 to make sure we keep the discussion in one place?
